src/hypothesis/extra/numpy.py

```diff
@@ -53,19 +53,24 @@ def from_dtype(dtype):
     elif dtype.kind == u'c':
         result = st.complex_numbers()
     elif dtype.kind in (u'S', u'a'):
-        result = st.binary()
+        result = st.binary(max_size=dtype.itemsize or None)
```

Comment:
Does this have an off-by-one error? If numpy can store variable size byte strings then it must either be length prefixing or zero terminating them. I'd assume the latter. But in that case itemsize must be counting the zero byte at the end. There's also a question of what happens if we have internal zeros here.

Reply:
The dtype for a variable-length string (unicode or bytestring) actually creates an array of fixed-length strings, but the limiting length is that of the longest string present when the array is initialised.

So the array is not variable size, but it's safe to generate any length string because the array doesn't exist yet!

Reply:
So what happens to the shorter strings? Do they get zero padded to the same length?

Reply:
Internally, I assume so, but the python API gives you `np.array(['aaaa', 'bb'])[1] == 'bb'` - the same strings you put in.   However assigning longer strings gets them truncated:  `np.array(['aaaa', 'bb'], 'U2')[0] == 'aa'`; that's the shrinking issue I mentioned.